pkg/operator/admission/webhook.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"github.com/pkg/errors"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/ceph/csi"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"os"
+)
+
+const (
+	appName            = "rook-admission-controller"
+	secretVolumeName   = "webhook-certificates"
+	serviceAccountName = "rook-admission-controller"
+	storageBackendCeph = "ceph"
+	portName           = "webhook-api"
+)
+
+var (
+	validatePath       = "/validate"
+	servicePort  int32 = 443
+	serverPort   int32 = 8079
+	namespace          = os.Getenv(k8sutil.PodNamespaceEnvVar)
+)
+
+func isSecretPresent(context *clusterd.Context) (bool, error) {
+	_, err := context.Clientset.CoreV1().Secrets(namespace).Get(appName, metav1.GetOptions{})
+	if err != nil {
+		// If secret is not found. All good ! Proceed with rook without admission controllers
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func createWebhookService(context *clusterd.Context) error {
```

Comment:
We really need some unit tests for the new admission package. These functions that create service, deployment, or pod specs are generally simple to test. You can see other operator unit tests for examples where services or other resources are created.

Reply:
Integration tests will also confirm the admission controller can start and run successfully. I believe we could keep this fairly simple and get a lot of validation with these steps:
- In the [CreateCephOperator()](https://github.com/rook/rook/blob/d8949ec945b036172588f587698e45633d5c392a/tests/framework/installer/ceph_installer.go#L81) method:
   - Create the RBAC. The easiest thing might be to add the yaml to GetRookOperator() in the ceph_manifests.go.
   - Call out to the script that generates the self-signed cert
- After the tests start the operator around [this line](https://github.com/rook/rook/blob/d8949ec945b036172588f587698e45633d5c392a/tests/framework/installer/ceph_installer.go#L426), verify that the admission controller is in running state
```
	if !h.k8shelper.IsPodInExpectedState("rook-ceph-admission-controller", onamespace, "Running") {
		logger.Error("admission controller is not running")
		return false, err
	}
```

Reply:
As a follow-up PR, please also add unit tests in addition to the integration tests.